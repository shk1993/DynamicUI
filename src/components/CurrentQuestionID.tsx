import * as React from 'react';

/**
 * This shows the current question the user is on above the submit buttons, for debugging purposes
 * @param props Current question object to display at the bottom fo the page
 * @returns     React component displaying the current questions
 */
export default (props: { currentQuestion: string[] }) => (
  <span style={{ color: 'white' }}>[{props.currentQuestion}]</span>
);
