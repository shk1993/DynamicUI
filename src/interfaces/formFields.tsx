import { AnyObject } from 'final-form';
import { FieldRenderProps } from 'react-final-form';
import { IAnswer } from './answers';
import { IAttribute } from './attributes';

/**
 * Input props for each of the form field elements
 */
export interface IInputProps {
  /** name of the question as a string */
  name: string;

  /** placeholder text for textboxes */
  placeholder: string;

  /** label to show the user */
  label: string;

  /** current values */
  values: AnyObject;

  /** answer choices */
  answers: IAnswer[];

  /** array of key and value strings for modifying the component */
  attributes?: IAttribute[] | undefined;

  /** push function used for field arrays */
  // tslint:disable-next-line:ban-types
  push?: Function;
}

/**
 * Props used for checkbox and radio components
 */
export interface IAnswerProps {
  /** label to show the user */
  label: string;

  /** current values */
  values: AnyObject;

  /** if exists, render the child component if the input is selected */
  child?: string;

  /** push function used for field arrays */
  // tslint:disable-next-line:ban-types
  push?: Function;
}

/**
 * Props used for the reusable component
 */
export interface IReusable extends IInputProps {
  /** type of component (checkbox or radio) as a string */
  type: string;

  /** component value */
  component:
    | React.ComponentType<FieldRenderProps>
    | React.StatelessComponent<FieldRenderProps>;
}
