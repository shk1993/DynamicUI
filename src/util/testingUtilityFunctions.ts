export const silenceConsoleErrors = () => {
  /* silence -- Warning: Stateless function components cannot be given refs. Attempts to access this ref will fail. */
  /* tslint:disable-next-line */
  console.error = () => {};
  /* tslint:disable-next-line */
  console.warn = () => {};
};
