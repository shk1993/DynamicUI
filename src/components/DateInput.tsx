import TextField from '@material-ui/core/TextField';
import * as Moment from 'moment';
import * as React from 'react';
import { DatePickerProps } from '../interfaces/datePickerProp';

import 'react-datepicker/dist/react-datepicker.css';
import './DatePicker.css';

const dateTimeFormat = 'MM/DD/YYYY HH:mm';
const dateFormat = 'MM/DD/YYYY';

/**
 * Displays the DatePicker component inside a material-ui TextField
 * @param props DatePickerProps used to manipulate the DatePicker values
 * @returns     The react-date-picker component in a material-ui TextField
 */
export const DateInput = (props: DatePickerProps) => (
  <TextField
    {...props}
    {...props.input}
    value={props.input.value !== 'UNK' ? props.input.value : ''}
    onBlur={event =>
      Moment(
        event.target.value,
        props.showTimeSelect ? dateTimeFormat : dateFormat,
        true,
      ).isValid()
        ? props.input.onChange(event.target.value)
        : props.input.onChange(undefined)
    }
    fullWidth
    error={props.meta.touched && !!props.meta.error}
    helperText={props.meta.touched && props.meta.error}
  />
);
