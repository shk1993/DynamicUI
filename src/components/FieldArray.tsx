import Button from '@material-ui/core/Button';
import Grid from '@material-ui/core/Grid';
import * as React from 'react';
import { FieldArray as FA } from 'react-final-form-arrays';
import { InputField } from '../components/FormFields';
import { IInputProps } from '../interfaces/formFields';
import QuestionSet from '../util/SampleQuestionSet';

/**
 * FieldArray is used for handling simple loops in the application. It initially displays a
 * simple ADD button, and if the user clicks it, the component will render all of it's children.
 * Clicking the ADD button again will render the children a second time, with different values in
 * the input components. This allows the user to input multiple values for a single question, used in loops.
 * @param props InputValues used in the input components
 * @returns     Our implementation of the FieldArray component
 */
export const FieldArray = (props: IInputProps) => {
  if (!props.attributes) {
    throw new Error('FieldArray MUST take attributes as a prop');
  }
  const childrenAttr = props.attributes.filter(
    attr => attr.key === 'children',
  )[0].value;
  const children = childrenAttr
    .substring(1, childrenAttr.length - 1)
    .split(',');
  return (
    <FA name={props.name}>
      {({ fields }) => (
        <div>
          {fields.map((LE, index) => (
            <Grid container key={LE}>
              {children.map(child => (
                <Grid item xs={6} key={child}>
                  <InputField
                    name={`${LE}.${QuestionSet[child].key}`}
                    placeholder={QuestionSet[child].text}
                    label={QuestionSet[child].text}
                    answers={[
                      {
                        key: 'noKey',
                        text: '',
                        nextQuestion: '',
                      },
                    ]}
                    values={props.values}
                  />
                </Grid>
              ))}
              <Button
                size="small"
                color="secondary"
                onClick={() => fields.remove(index)}
              >
                Remove
              </Button>
            </Grid>
          ))}
          <Button
            onClick={() => props.push && props.push('specs', undefined)}
            size="medium"
            color="primary"
          >
            Add
          </Button>
        </div>
      )}
    </FA>
  );
};
