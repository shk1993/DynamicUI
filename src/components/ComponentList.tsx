import { FieldArray } from '../components/FieldArray';
import {
  CheckboxField,
  DatePicker,
  InputField,
  RadioField,
} from '../components/FormFields';
import Review from '../pages/Review';

export default {
  InputField,
  RadioField,
  CheckboxField,
  DatePicker,
  Review,
  FieldArray,
};
