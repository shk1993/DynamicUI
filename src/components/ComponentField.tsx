import Typography from '@material-ui/core/Typography';
import * as React from 'react';
import ComponentList from '../components/ComponentList';
import ICurrentQuestion from '../interfaces/currentQuestion';
import { IRenderComponent } from '../interfaces/parseRender';
import { setHistory } from '../state/historyState';
import FlattenAnswers from '../util/FlattenAnswers';
import ParseRender from '../util/ParseRender';
import ReturnAnswers from '../util/ReturnAnswers';
import QuestionSet from '../util/SampleQuestionSet';

/**
 * The starting point of the main purpose of our application and called from the QuestionSet page.
 * @param props The current question, history values, and the push function
 * @returns     React component consisting of the question(s) and answer(s)
 */
export default (props: ICurrentQuestion) => {
  const renderGroup = [];
  setHistory(props.values);
  for (const question of props.currentQuestion) {
    const curQuestion = QuestionSet[question];
    const render = ParseRender(curQuestion.render);
    const flatAnswers = ReturnAnswers(
      render,
      FlattenAnswers(curQuestion.answers),
    );
    renderGroup.push({ question: curQuestion, answers: flatAnswers, render });
  }
  return (
    <div>
      {renderGroup.map(groupComp => {
        return (
          <div key={JSON.stringify(groupComp)}>
            <Typography variant="subheading">
              {groupComp.question.text}
            </Typography>
            {groupComp.render.map((comp: IRenderComponent) => {
              const Component = ComponentList[comp.component];
              return (
                <Component
                  key={JSON.stringify(comp)}
                  label={groupComp.question.text}
                  name={groupComp.question.key}
                  answers={groupComp.answers}
                  values={props.values}
                  attributes={comp.attributes}
                  push={props.push}
                />
              );
            })}
            <br />
          </div>
        );
      })}
    </div>
  );
};
