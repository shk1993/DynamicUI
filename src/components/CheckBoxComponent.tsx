import Checkbox from '@material-ui/core/Checkbox';
import * as React from 'react';
import { FieldRenderProps } from 'react-final-form';
import { IAnswerProps } from '../interfaces/formFields';
import ComponentField from './ComponentField';

/**
 * Displays the checkbox element by material-ui. If we have a child, and if the checkbox is selected, then display the child component
 * @param props react-final-form's field props and custom props we need for checkboxes
 * @returns     checkbox components, showing the label and the child if selected
 */
export const CheckBoxElement = (props: FieldRenderProps & IAnswerProps) => (
  <div>
    <Checkbox
      {...props}
      {...props.input}
      name={props.input.name}
      onChange={props.input.onChange}
      value={props.input.value}
    />
    {'\n'}
    {props.label}
    <br />
    {props.child &&
      props.input.checked && (
        <ComponentField currentQuestion={[props.child]} values={props.values} />
      )}
  </div>
);
