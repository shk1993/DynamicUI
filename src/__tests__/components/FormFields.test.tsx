import * as React from 'react';
import { render } from 'react-testing-library';

import { Stack } from 'immutable';
import { FieldArray } from '../../components/FieldArray';
import {
  CheckboxField,
  DatePicker,
  InputField,
  RadioField,
} from '../../components/FormFields';
import { IInputProps } from '../../interfaces/formFields';
import { silenceConsoleErrors } from '../../util/testingUtilityFunctions';

Date.now = jest.fn(() => new Date('2017-01-01T00:00:00+0000'));

beforeAll(() => silenceConsoleErrors());

describe('Test FormField component', () => {
  const inputProps: IInputProps = {
    label:
      'What is the date the Common Formats patient safety module was completed?',
    name: 'DE30',
    placeholder: 'dd',
    answers: [{ key: 'noKey', text: '', nextQuestion: 'DE3' }],
    attributes: [{ key: 'children', value: '[DE112,DE113]' }],
    push: jest.fn(),
    values: {},
  };

  it('checks whether the DatePicker matches snapshot', () => {
    const { container } = render(<DatePicker {...inputProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('checks whether the RadioField matches snapshot', () => {
    const { container } = render(<RadioField {...inputProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('checks whether the InputField matches snapshot', () => {
    const { container } = render(<InputField {...inputProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('checks whether the CheckboxField matches snapshot', () => {
    const { container } = render(<CheckboxField {...inputProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('checks whether the FieldArray matches snapshot', () => {
    const { container } = render(<FieldArray {...inputProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
