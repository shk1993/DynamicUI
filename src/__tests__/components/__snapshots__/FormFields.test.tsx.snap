// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test FormField component checks whether the CheckboxField matches snapshot 1`] = `
<div>
  <label>
    <div>
      <span
        class="MuiButtonBase-root-43 MuiIconButton-root-37 MuiSwitchBase-root-33 MuiCheckbox-root-62 MuiCheckbox-colorSecondary-66"
        input="[object Object]"
        label=""
        meta="[object Object]"
        values="[object Object]"
      >
        <span
          class="MuiIconButton-label-42"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-46"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
          <input
            class="MuiSwitchBase-input-36"
            name="DE30"
            type="checkbox"
            value="noKey"
          />
        </span>
        <span
          class="MuiTouchRipple-root-52"
        />
      </span>
      

      
      <br />
    </div>
  </label>
</div>
`;

exports[`Test FormField component checks whether the DatePicker matches snapshot 1`] = `
<div>
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container"
    >
      <div
        answers="[object Object]"
        attributes="[object Object]"
        class="MuiFormControl-root-1 MuiFormControl-fullWidth-4"
        input="[object Object]"
        meta="[object Object]"
        values="[object Object]"
      >
        <label
          class="MuiFormLabel-root-10 MuiInputLabel-root-5 MuiInputLabel-formControl-6 MuiInputLabel-animated-9"
          data-shrink="false"
        >
          dd
        </label>
        <div
          class="MuiInput-root-15 MuiInput-fullWidth-22 MuiInput-formControl-16 MuiInput-underline-19"
        >
          <input
            aria-invalid="false"
            aria-required="false"
            class="MuiInput-input-23"
            name="DE30"
            placeholder="MM/DD/YYYY"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Test FormField component checks whether the FieldArray matches snapshot 1`] = `
<div>
  <button
    class="MuiButtonBase-root-43 MuiButton-root-67 MuiButton-flatPrimary-69"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label-68"
    >
      Add
    </span>
    <span
      class="MuiTouchRipple-root-52"
    />
  </button>
</div>
`;

exports[`Test FormField component checks whether the InputField matches snapshot 1`] = `
<div
  answers="[object Object]"
  attributes="[object Object]"
  class="MuiFormControl-root-1 MuiFormControl-fullWidth-4"
  input="[object Object]"
  meta="[object Object]"
  values="[object Object]"
>
  <label
    class="MuiFormLabel-root-10 MuiInputLabel-root-5 MuiInputLabel-formControl-6 MuiInputLabel-animated-9"
    data-shrink="false"
  >
    dd
  </label>
  <div
    class="MuiInput-root-15 MuiInput-fullWidth-22 MuiInput-formControl-16 MuiInput-multiline-21 MuiInput-underline-19"
  >
    <div
      class="Textarea-root-59"
      style="height: 0px;"
    >
      <textarea
        aria-hidden="true"
        class="Textarea-shadow-61 Textarea-textarea-60"
        readonly=""
        rows="1"
        tabindex="-1"
      />
      <textarea
        aria-hidden="true"
        class="Textarea-shadow-61 Textarea-textarea-60"
        readonly=""
        rows="1"
        tabindex="-1"
      />
      <textarea
        aria-invalid="false"
        aria-required="false"
        class="Textarea-textarea-60 MuiInput-input-23 MuiInput-inputMultiline-25"
        name="DE30"
        placeholder="dd"
        rows="1"
        type="text"
      />
    </div>
  </div>
</div>
`;

exports[`Test FormField component checks whether the RadioField matches snapshot 1`] = `
<div>
  <label>
    <div>
      <span
        class="MuiButtonBase-root-43 MuiIconButton-root-37 MuiSwitchBase-root-33 MuiRadio-root-28 MuiRadio-colorSecondary-32"
        input="[object Object]"
        label=""
        meta="[object Object]"
        values="[object Object]"
      >
        <span
          class="MuiIconButton-label-42"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root-46"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
            />
          </svg>
          <input
            class="MuiSwitchBase-input-36"
            name="DE30"
            type="radio"
            value="noKey"
          />
        </span>
        <span
          class="MuiTouchRipple-root-52"
        />
      </span>
      

      
      <br />
    </div>
  </label>
</div>
`;
